### PR TITLE
Be resilient to options being serialized as strings.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @pilchie

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @pilchie

File: src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs [hunk 1]
```diff
@@ -96,6 +95,27 @@ bool IOptionPersister.TryFetch(OptionKey optionKey, out object value)
                         value = null;
                         return false;
                     }
+                    else if (optionKey.Option.Type == typeof(int))
+                    {
+                        var untypedValue = subKey.GetValue(key, defaultValue: optionKey.Option.DefaultValue);
+
+                        if (untypedValue is string stringValue)
+                        {
+                            // Due to a previous bug we were accidentally serializing longs as strings. Gracefully convert
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs]: ❓ How does this apply to `int` values?

File: src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs [hunk 2]
```diff
@@ -77,24 +77,41 @@ bool IOptionPersister.TryFetch(OptionKey optionKey, out object value)
...
+
+                            case long longValue:
+                                value = longValue;
+                                return true;
                         }
-                        else if (untypedValue is long longValue)
+                    }
+                    else if (optionKey.Option.Type == typeof(int))
+                    {
+                        var untypedValue = subKey.GetValue(key, defaultValue: optionKey.Option.DefaultValue);
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs]: ❔ Why not use `Convert.ToInt32(untypedValue)`?

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Options/LocalUserRegistryOptionPersister.cs]: I think we want to be super crisp about how things convert to/from registry form.  And, in the case where we've stored data in the wrong form, we want to be crisp about the code that gracefully handles it.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski Despite your fix, i still got this (presumably due to some bad cached values locally on my box.  So i expanded your fix to not just handle 'longs' but also 'ints' as well.

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 

**Customer scenario**

The recent change to add hte goldbar to offer the new 'fxcop vsix experience' ended up serializing some data badly into the registry.  This change makes it so we gracefully handle that case for machines that may have run that code in the interim period before we fixed that issue.

**Bugs this fixes:**

None.

**Workarounds, if any**

Delete specific registry keys.  No way to know easily know which you'd have to delete.  Until you fix it VS just crashes.

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

The original code to store 'gold bar' data was tested locally without actually shutting down VS.  So we never exercised the codepaths that actually persisted to the registry and back.  Instead, we only really tested in memory representation.  The registry codepaths were not correct, and you could get a failure to deserialize the data.  We've fixed this up, but machines that participated in this small window of time may still have bad data serialized.  This makes us resilient ot that so that these users will not crash.

**How was the bug found?**

Local dogfooding.

**Pull request merged**